### ISSUE TITLE
[Bug]: ReadAllFiles does not fully read gzipped files from GCS

### ISSUE THREAD

**Issue by janowskijak** [apache/beam#31040]
### What happened?

Since the refactor of gcsio (2.52?) ReadAllFiles does not fully read gzipped files from GCS. Part of the file will be correctly returned but rest will go missing. 

I presume this is caused by the fact that GCS performs [decompressive transcoding](https://cloud.google.com/storage/docs/transcoding#decompressive_transcoding) while `_ExpandIntoRanges` uses the GCS objects metadata to determine the read range. This means that the file size we receive is larger than the maximum of the read range. 

For example, a gzip on GCS might have a file size of 1 MB and this will be the object size in the metadata. Thus the maximum of the read range will be 1 MB. However, when beam opens the file it's already decompressed by GCS so the file size will be 1.5 MB and we won't read 0.5 MB out of it thus causing data loss.

### Issue Priority

Priority: 1 (data loss / total loss of function)

### Issue Components

- [X] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

**Comment by Abacn** [comment]
Thanks for reporting. Agree this is a P1 bug as it causes data loss.

**Comment by Abacn** [comment]
Is it possible to provide a working example that reproduce the issue, which could help triage.

**Comment by liferoad** [comment]
@shunping FYI

**Comment by janowskijak** [comment]
> Is it possible to provide a working example that reproduce the issue, which could help triage.

@Abacn I don't have a working example however the steps to reproduce are:

1. Upload a gzip file to GCS. Make sure that the unzipped file is large enough, e.g a few MB. 
2. Create a beam pipeline using Python SDK that reads the file from 1. using RealAllFromText.
3. Print or write the output of ReadAllFromText.
4. Observe that the file is not fully read.

**Comment by Michal-Nguyen-airspace-intelligence** [comment]
So I double checked and there are differences between your example and our case.

- We use content encoding `gzip` while saving our files to GCS, you don't have encoding specified
- This leads us to using `ReadAllFromText` with parameter `compression_type=CompressionTypes.UNCOMPRESSED` since the downloaded file seems to be already uncompressed (it doesn't work with `CompressionTypes.AUTO`), as in gcs policy
- This further results in reading only fragment of the file

Furthermore, after removing encoding type from our file and using `CompressionTypes.AUTO` on it worked properly.
To get you example to represent our situation please add content encoding `gzip` to your file metadata.

**Comment by Michal-Nguyen-airspace-intelligence** [comment]
For quick patch we use following solution:
```
class ReadAllFromTextNotSplittable(ReadAllFromText):
    """This class doesn't take advantage of splitting files in bundles because
    when doing so beam was taking compressed file size as reference resulting in
    reading only a fracture of uncompressed file"""

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._read_all_files._splittable = False
```

**Comment by liferoad** [comment]
What does your medadata look like?

I tried this:
<img width="538" alt="image" src="https://github.com/apache/beam/assets/7833268/f96672a8-b2bc-4f85-a1fe-45ecd0d03ab2">

Then I got this error:
```
ERROR:apache_beam.runners.common:Error -3 while decompressing data: incorrect header check [while running '[6]: Read File from GCS/ReadAllFiles/ReadRange']
```

**Comment by Michal-Nguyen-airspace-intelligence** [comment]
This is expected, as I mentioned earlier 
```This leads us to using ReadAllFromText with parameter compression_type=CompressionTypes.UNCOMPRESSED since the downloaded file seems to be already uncompressed (it doesn't work with CompressionTypes.AUTO), as in gcs policy```
I presume while downloading file from GCS it's already decompressed, hence the error of decompression in Beam.

**Comment by liferoad** [comment]
I see. 

```
# standard libraries
import logging

# third party libraries
import apache_beam as beam
from apache_beam import Create, Map
from apache_beam.io.textio import ReadAllFromText
from apache_beam.options.pipeline_options import PipelineOptions
from apache_beam.transforms.combiners import Count

logger = logging.getLogger()
logger.setLevel(logging.INFO)

elements = [
    # "gs://apache-beam-samples/gcs/bigfile.txt.gz",
    # "gs://apache-beam-samples/gcs/bigfile_with_encoding.txt.gz",
    "gs://apache-beam-samples/gcs/bigfile_with_encoding_plain.txt.gz",
]

options = PipelineOptions()

with beam.Pipeline(options=options) as p:
    (
        p
        | Create(elements)
        | "Read File from GCS"
        >> ReadAllFromText(
            compression_type=beam.io.filesystem.CompressionTypes.UNCOMPRESSED
        )
        | Count.Globally()
        | "Log" >> Map(lambda x: logging.info("Total lines %d", x))
    )
```
This only loads 75,601 lines.

#19413 could be related for uploading the file to GCS.

**Comment by shunping** [comment]
.take-issue

**Comment by kennknowles** [comment]
Have we reproduced this?

**Comment by liferoad** [comment]
Yes, see my above link: https://github.com/apache/beam/issues/31040#issuecomment-2081172626

**Comment by kennknowles** [comment]
Is there a hope of a fix for 2.57.0 cherry pick? I would guess this is a longstanding issue so getting it fixed in a very thorough way for 2.58.0 is actually the best thing to do. I recall we had decompressive transcoding bugs in the past. So we should make sure we really get it right this time. And the user can mitigate by configuring GCS to not do the transcoding.

**Comment by liferoad** [comment]
Moved this to 2.58.0. Thanks!

**Comment by jrmccluskey** [comment]
Has any progress been made on this?

**Comment by lostluck** [comment]
Has any progress been made on this?

**Comment by lostluck** [comment]
Moved to 2.60.0

**Comment by serratedserenade** [comment]
Has there been any progress on this? This is starting to become a blocker for us.

**Comment by liferoad** [comment]
https://github.com/apache/beam/issues/31040#issuecomment-2072118260 does not work for you?

**Comment by serratedserenade** [comment]
> [#31040 (comment)](https://github.com/apache/beam/issues/31040#issuecomment-2072118260) does not work for you?

That's what I was using prior until we realised this method was causing part of the data we were reading to be skipped. Basically same problem as here https://github.com/apache/beam/issues/31040#issuecomment-2081172626

**Comment by kennknowles** [comment]
This issue came up and was fixed in the Java SDK file IO many years ago, so we should have some reference material to work with. I am trying to find it.

And presumably also related to #18390.

**Comment by shunping** [comment]
Pick up this issue again. I confirm that I am able to reproduce the problem with code in https://github.com/apache/beam/issues/31040#issuecomment-2081172626

**Comment by shunping** [comment]
I briefly debugged the issue, it seems that file is accessed with decompressive transcoding  (https://cloud.google.com/storage/docs/transcoding#decompressive_transcoding).

I checked the metadata of the GCS file in use, and the result is as follows.
```bash
$ gcloud storage objects describe gs://apache-beam-samples/gcs/bigfile_with_encoding_plain.txt.gz
acl:
...
bucket: apache-beam-samples
content_encoding: gzip
content_type: text/plain
crc32c_hash: /REfiQ==
creation_time: 2024-04-27T20:19:48+0000
...
metageneration: 1
name: gcs/bigfile_with_encoding_plain.txt.gz
size: 7635685
storage_class: STANDARD
storage_class_update_time: 2024-04-27T20:19:48+0000
storage_url: gs://apache-beam-samples/gcs/bigfile_with_encoding_plain.txt.gz#1714249188497359
update_time: 2024-04-27T20:19:48+0000
```

The size field shows the size of the gzip file. The original file size prior to gzip should be 10100000.
```
$ wc bigfile_with_encoding_plain.txt
  100000  252788 10100000 bigfile_with_encoding_plain.txt
```

**Comment by shunping** [comment]
Just check around. It seems GCS decompressive transcoding causes some similar issues elsewhere.
- https://github.com/piskvorky/smart_open/issues/422
- https://github.com/apache/opendal/issues/5070

**Comment by shunping** [comment]
The range is determined by the size obtained from GCSIO (https://github.com/apache/beam/blob/d7502fa73b153a81d73bf5d658e1b6a7ae0747c6/sdks/python/apache_beam/io/gcp/gcsfilesystem.py#L339, https://github.com/apache/beam/blob/d7502fa73b153a81d73bf5d658e1b6a7ae0747c6/sdks/python/apache_beam/io/gcp/gcsio.py#L477), which in turn called GCS Python Client Library `google-cloud-storage`. 

The size property in the Blob object there comes from "Content-Length". (https://cloud.google.com/storage/docs/json_api/v1/objects) 
![image](https://github.com/user-attachments/assets/6fa0b60a-5cad-4442-81c4-ddf0f7bbc892)

**Comment by shunping** [comment]
To clarify the behavior of textio with various content encoding, content type, and compression settings, I've expanded the table in the Apache Beam GitHub issue [#18390](https://github.com/apache/beam/issues/18390#issuecomment-1422729486). This table compares the behavior across two Beam SDK versions: 2.52.0 (prior to the GCSIO migration) and 2.62.0 (the upcoming release). I also include the proposed behavior of my fix in the last column.

![image](https://github.com/user-attachments/assets/629a95f4-d72c-47ce-9ab6-8d7e180fa3d2)

A few notes about how the data is generated.
- For the first 3 x 2 x 3 rows, the text data is gzipped locally and then uploaded to gcs. Then the metadata values of `content-type` and `content-encoded` are manually adjusted.
- For the row marked as "copy default text file", the text data is directly copied/uploaded to gcs without gzip.
- For the row marked as "copy default gzip file", the gzipped text data is copied/uploaded to gcs.
- For the row marked as "copy default text file with gzip-local flag", the **text data** is uploaded to gcs with the said flag. 
 `gcloud storage cp -Z ./textio-test-data.1k.txt gs://apache-beam-samples/textio/textio-test-data.gzip-local.1k.txt.gz`.

**Comment by shunping** [comment]
Notice that when `content-type=application/gzip` and `content-encoding=gzip`, GCS considers the file is doubly compressed(https://cloud.google.com/storage/docs/transcoding#gzip-gzip), which is against the actual content we store here (aka. a gzipped text file). Therefore, an exception is thrown there.

For any `UnicodeDecodeError`, most of them are due to the fact that users specifies "UNCOMPRESSED" on a gzip file. We can let it be this way, or we can give a better and more informative error if "content-type" and/or "content-encoding" does not match the specified "compression type". 

@kennknowles WDYT?

**Comment by kennknowles** [comment]
I agree with all of your proposals that replace "Data Loss" with "UnicodeDecodeError"

**Comment by kennknowles** [comment]
In all the cases where GCS transcoding causes a zlib.error I also agree with making them function correctly. Basically if the user says it is GZIP and the data really is GZIP but we know that GCS is going to decode it then we do not do a redundant decode.

**Comment by shunping** [comment]
> I agree with all of your proposals that replace "Data Loss" with "UnicodeDecodeError"

Great! GCS decompressive transcoding is a bit unintuitive to Beam users here, and when it happens, we see data loss. I think it is more natural to expect users to specify GZIP or AUTO in those cases rather than UNCOMPRESSED, as shown in the proposal.

> Basically if the user says it is GZIP and the data really is GZIP but we know that GCS is going to decode it then we do not do a redundant decode.

Yep, that's the idea, but it is implemented it differently in the proposed fix (#33384). 

- Instead of trying to determine whether GCS is going to do decompressive transcoding, which is both unclear and inconvenient to verify,  we can call GCS client library and let it always return raw data without transcoding. 
- At first, I was worrying about performance of this approach. As the gzip decoding will then happen on our side (Beam), which is different from server-side decoding that is mentioned in https://cloud.google.com/storage/docs/transcoding.
- However, after closely examining the GCS client library, I discover that the GCS client library actually **always** requests GCS for raw data. It then adds an extra decoding step in itself (https://github.com/googleapis/google-resumable-media-python/blob/402feb7b38a972daad9bd3e26b80ddd0879bd53f/google/resumable_media/requests/download.py#L127) before returning that to the caller (in this case Beam). It merely mimicks the effects of server-side decompressive transcoding, but the actual decoding workload is done on the client.

In my fix (#33384), I let GCS client library skip the decoding step in itself and rely on our own decoding mechanism in CompressedFile to process the file. It works exactly as what has been proposed in the previous table.
